### DISCUSSION TITLE
How to get Android equivalent of `CFBundleVersion` and other `Bundle.infoDictionary` values?

### DISCUSSION THREAD

**Discussion by dbystruev** [skiptools/skip#223] [Q&A] [answered]
I have the following code in Swift:
```Swift
#if !SKIP
extension Bundle {
    var appBuild: String { getInfo("CFBundleVersion") }
    var appVersionLong: String { getInfo("CFBundleShortVersionString") }
    var displayName: String { getInfo("CFBundleDisplayName") }
    func getInfo(_ string: String) -> String { infoDictionary?[string] as? String ?? "⚠️" }
}
#endif
```

It is used as the following:
```Swift
var version: String {
#if SKIP
        "Android"
#else
        "\(Bundle.main.displayName) v\(Bundle.main.appVersionLong) build \(Bundle.main.appBuild)"
#endif
    }
}
```

Is there any way to get access to equivalent of `Bundle.infoDictionary` for Android code?

**Comment by marcprux** [comment, marked as answer]
There's no exact equivalent for `Bundle.infoDictionary` on Android – the closest thing is to use the [android.content.pm.PackageInfo](https://developer.android.com/reference/android/content/pm/PackageInfo) class to inspect the app's metadata. You can see how the [Skip Showcase](https://github.com/skiptools/skipapp-showcase/blob/main/Sources/Showcase/AboutView.swift#L91) app implements the equivalent of `CFBundleShortVersionString`:

```swift
let appVersion: String? = {
    #if !SKIP
    return Bundle.main.infoDictionary?["CFBundleShortVersionString"] as? String
    #else
    let context = ProcessInfo.processInfo.androidContext
    let packageManager = context.getPackageManager()
    let packageInfo = packageManager.getPackageInfo(context.getPackageName(), android.content.pm.PackageManager.GET_META_DATA)
    let versionName = packageInfo.versionName
    let versionCode = packageInfo.versionCode
    return versionName
    #endif
}()
```

Looking at the package info docs, it looks like you could also do the equivalent for `CFBundleDisplayName` with the [packageName](https://developer.android.com/reference/android/content/pm/PackageInfo#packageName) property.

This could be an interesting thing to bake into SkipFunction's [Bundle](https://github.com/skiptools/skip-foundation/blob/e459ad7b32828d0f4aa0fe6d0c764566fb0e0138/Sources/SkipFoundation/Bundle.swift#L404) implementation – checking whether the bundle `.main` and synthesizing the various standard properties from the equivalent manifest entires.

**Reply by dbystruev** [reply]
Great, that works! Also it is clear now how to get access to other variables available from `getPackageInfo`.
